### PR TITLE
Download student projects

### PR DESCRIPTION
**Pull request by AndreyNikitin**
src/test/download.ts

### CODE REVIEW

File: src/lib/course.ts [hunk 1]
```diff
@@ -71,9 +71,79 @@ export async function assignmentStudentsProgress(courseId: string, assignmentId:
...
+    const domain = config.getDomain()
+    const authHeaders = {
+      'Authorization': `Bearer ${token}`
+    }
+    const archive = await getJson(`https://octopus.${domain}/api/v1/courses/${courseId}/assignments/${assignmentId}/students/${studentId}/download/${taskId}`, undefined, authHeaders)
+    if (!archive.done) {
+      await new Promise(resolve => setTimeout(resolve, 500))
+      return await waitDownloadTask(courseId, assignmentId, studentId, taskId)
+    } else {
+      return archive.url
```
Comment (destitutus) [inline src/lib/course.ts:90]: just return without else branch

File: src/lib/course.ts [hunk 2]
```diff
@@ -71,9 +71,79 @@ export async function assignmentStudentsProgress(courseId: string, assignmentId:
...
+    const token = config.getToken()
+    const domain = config.getDomain()
+    const authHeaders = {
+      'Authorization': `Bearer ${token}`
+    }
+    const taskId = await getJson(`https://octopus.${domain}/api/v1/courses/${courseId}/assignments/${assignmentId}/students/${studentId}/download`, undefined, authHeaders)
+    try {
+      // if not string then it is error object
+      const jsonRes = JSON.parse(taskId)
+      throw new Error(JSON.stringify(jsonRes))
```
Comment (destitutus) [inline src/lib/course.ts:116]: not sure if I understand the logic - we throw a error and after that we catch all error in silent way

File: src/lib/course.ts [hunk 3]
```diff
@@ -71,9 +71,79 @@ export async function assignmentStudentsProgress(courseId: string, assignmentId:
...
+      throw new Error(message)
+    }
+    throw error
+  }
+}
+
+export async function downloadStudentsAssignments(courseId: string, assignmentId: string, filter: (s: StudentProgress) => boolean): Promise<string[]> {
+  let students = await assignmentStudentsProgress(courseId, assignmentId)
+  students = students.filter(filter)
+  console.log('students', students)
```
Comment (destitutus) [inline src/lib/course.ts:134]: no need console.log here more I guess

File: src/test/download.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        date.setUTCHours(0, 0, 0, 0)
+        const links = await codio.v1.course.downloadStudentsAssignments(courseId, assignmentId, (studentProgress) => {
+            return studentProgress.completion_date > date
+        })
+        console.log('links', links)
+        await Promise.all(links.map(link => {
+            const parsed = path.parse(link)
+            const dest = `${folder}/${parsed.name}${parsed.ext}`
+            return downloadFile(link, dest)
+                .then(() => console.log('downloaded', link))
```
Comment (destitutus) [inline src/test/download.ts:51]: do we need cleanup after test end?

**Review by destitutus — APPROVED** [review]

**Review by vvlevykin — APPROVED** [review]

**Pull request merged**